File: internal/webui/webauthn.go
package webui

import (
	"encoding/base64"
	"encoding/json"
	"io"
	"net/http"

	"github.com/go-webauthn/webauthn/protocol"

	"github.com/archer-developer/miranda/internal/session"
)

// handleWebAuthnRegisterBegin starts an "add a passkey" ceremony for the
// logged-in user, from the profile screen. Keyed by the caller's own
// session token — registration only ever happens once already
// authenticated, so there's already a stable per-request key to stash the
// pending ceremony under; no separate ceremony id needs to round-trip to
// the client the way login's does (see handleWebAuthnLoginBegin).
func (h *Handler) handleWebAuthnRegisterBegin(w http.ResponseWriter, r *http.Request) {
	user := currentUser(r)
	cookie, err := r.Cookie(session.CookieName)
	if err != nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	creation, err := h.webauthn.BeginRegistration(r.Context(), user.Username, cookie.Value)
	if err != nil {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}
	writeJSON(w, creation)
}

// webauthnRegisterFinishRequest is the browser's PublicKeyCredential JSON
// from navigator.credentials.create(), plus one sibling field this
// endpoint needs and the WebAuthn spec doesn't define. Only Nickname is
// read here — the rest of the body is re-read whole by
// webauthn.Service.FinishRegistration, which needs the full standard
// shape, not just this one field.
type webauthnRegisterFinishRequest struct {
	Nickname string `json:"nickname"`
}

// handleWebAuthnRegisterFinish completes a passkey registration ceremony.
func (h *Handler) handleWebAuthnRegisterFinish(w http.ResponseWriter, r *http.Request) {
	user := currentUser(r)
	cookie, err := r.Cookie(session.CookieName)
	if err != nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "failed to read request body", http.StatusBadRequest)
		return
	}
	var req webauthnRegisterFinishRequest
	_ = json.Unmarshal(body, &req) // nickname is optional; malformed JSON surfaces below anyway

	info, err := h.webauthn.FinishRegistration(r.Context(), user.Username, cookie.Value, req.Nickname, body)
	if err != nil {
		h.logger.Warn("webauthn: registration failed",
			"error", err, "username", user.Username, "remote_addr", r.RemoteAddr, "user_agent", r.UserAgent())
		http.Error(w, "registration failed: "+err.Error(), http.StatusBadRequest)
		return
	}
	writeJSON(w, info)
}

// handleWebAuthnLoginBegin starts a passwordless, usernameless login
// ceremony for the login page's biometric button. Unauthenticated by
// design — this *is* the login mechanism.
func (h *Handler) handleWebAuthnLoginBegin(w http.ResponseWriter, r *http.Request) {
	assertion, ceremonyID, err := h.webauthn.BeginDiscoverableLogin(r.Context())
	if err != nil {
		h.logger.Warn("webauthn: discoverable login begin failed",
			"error", err, "remote_addr", r.RemoteAddr, "user_agent", r.UserAgent())
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}
	writeJSON(w, webauthnLoginBeginResponse{CeremonyID: ceremonyID, CredentialAssertion: assertion})
}

// webauthnLoginBeginResponse embeds the standard WebAuthn assertion options
// at the top level (so the browser's navigator.credentials.get() can use
// the JSON as-is) alongside the one sibling field — the ceremony id — the
// client must echo back on finish, since there's no session cookie yet to
// key the pending challenge by.
type webauthnLoginBeginResponse struct {
	CeremonyID string `json:"ceremonyId"`
	*protocol.CredentialAssertion
}

// handleWebAuthnLoginFinish completes a passwordless login ceremony and, on
// success, issues a session cookie exactly like password login does (see
// auth.go's issueSessionCookie) — the two auth methods share that one code
// path so they can never drift apart.
func (h *Handler) handleWebAuthnLoginFinish(w http.ResponseWriter, r *http.Request) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "failed to read request body", http.StatusBadRequest)
		return
	}
	var req struct {
		CeremonyID string `json:"ceremonyId"`
	}
	if err := json.Unmarshal(body, &req); err != nil || req.CeremonyID == "" {
		http.Error(w, "missing ceremonyId", http.StatusBadRequest)
		return
	}

	username, credentialID, prfOutput, err := h.webauthn.FinishDiscoverableLogin(r.Context(), req.CeremonyID, body)
	if err != nil {
		// The client only ever shows a generic "biometric sign-in failed"
		// message (see login.js) — this is the only place the actual
		// failure reason (which credential, which validation step) is
		// visible at all, which matters a lot for a ceremony this hard to
		// reproduce from a desk (it depends on a specific phone's
		// authenticator/OS behavior, e.g. Android's Google Password
		// Manager — see Store.ReconcileFlags for one such quirk already
		// worked around).
		h.logger.Warn("webauthn: discoverable login failed",
			"error", err, "remote_addr", r.RemoteAddr, "user_agent", r.UserAgent())
		http.Error(w, "login failed: "+err.Error(), http.StatusUnauthorized)
		return
	}

	user, ok := h.users.Get(username)
	if !ok {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	// Best-effort, same reasoning as handleLoginSubmit's password path:
	// prfOutput is nil whenever the authenticator/browser doesn't support
	// PRF, which UnlockWithPRF treats as a normal no-op, not an error.
	if h.keyring != nil {
		if err := h.keyring.UnlockWithPRF(r.Context(), username, credentialID, prfOutput); err != nil {
			h.logger.Warn("keyring: unlock with prf failed", "username", username, "error", err)
		}
	}

	if err := h.issueSessionCookie(w, r, user); err != nil {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}
	writeJSON(w, map[string]string{"redirect": "/"})
}

// handleWebAuthnListCredentials lists the logged-in user's registered
// passkeys for the profile screen's "manage passkeys" section.
func (h *Handler) handleWebAuthnListCredentials(w http.ResponseWriter, r *http.Request) {
	user := currentUser(r)
	list, err := h.webauthn.ListCredentials(r.Context(), user.Username)
	if err != nil {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}
	writeJSON(w, list)
}

// handleWebAuthnDeleteCredential removes one of the logged-in user's own
// passkeys — Service.DeleteCredential scopes the delete by username, so
// this can never remove another account's credential even given a guessed id.
func (h *Handler) handleWebAuthnDeleteCredential(w http.ResponseWriter, r *http.Request) {
	user := currentUser(r)
	id, err := base64.RawURLEncoding.DecodeString(r.PathValue("id"))
	if err != nil {
		http.Error(w, "invalid credential id", http.StatusBadRequest)
		return
	}

	if err := h.webauthn.DeleteCredential(r.Context(), user.Username, id); err != nil {
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	// Best-effort: drop this credential's now-orphaned wrapped-key slot too,
	// if it had one. Never fails the delete itself — the passkey is already
	// gone from the auth flow at this point regardless.
	if h.keyring != nil {
		if err := h.keyring.RemoveSlotForCredential(r.Context(), user.Username, id); err != nil {
			h.logger.Warn("keyring: remove slot for deleted credential failed", "username", user.Username, "error", err)
		}
	}

	w.WriteHeader(http.StatusNoContent)
}

// webauthnProbeRequest carries which credential a probe ceremony targets —
// the request body for probe-begin, and one sibling field alongside the
// full navigator.credentials.get() response for probe-finish, same
// convention as webauthnRegisterFinishRequest's nickname field.
type webauthnProbeRequest struct {
	CredentialID string `json:"credentialId"` // base64url, matches CredentialInfo.ID
}

// parseProbeRequest reads and validates the request body shared by
// handleWebAuthnRegisterProbeBegin/Finish: the logged-in user, the session
// cookie (ceremonies are keyed by it, same as registration), the raw body
// (probe-finish needs it whole, to hand to FinishKeyProbe), and the decoded
// credentialId. Writes the appropriate HTTP error and returns ok=false on
// any failure — callers should return immediately in that case.
func (h *Handler) parseProbeRequest(w http.ResponseWriter, r *http.Request) (username, ceremonyKey string, body, credentialID []byte, ok bool) {
	user := currentUser(r)
	cookie, err := r.Cookie(session.CookieName)
	if err != nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return "", "", nil, nil, false
	}

	body, err = io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "failed to read request body", http.StatusBadRequest)
		return "", "", nil, nil, false
	}
	var req webauthnProbeRequest
	if err := json.Unmarshal(body, &req); err != nil || req.CredentialID == "" {
		http.Error(w, "missing credentialId", http.StatusBadRequest)
		return "", "", nil, nil, false
	}
	credentialID, err = base64.RawURLEncoding.DecodeString(req.CredentialID)
	if err != nil {
		http.Error(w, "invalid credentialId", http.StatusBadRequest)
		return "", "", nil, nil, false
	}
	return user.Username, cookie.Value, body, credentialID, true
}

// handleWebAuthnRegisterProbeBegin starts the follow-up assertion ceremony
// (see internal/webauthn.Service.BeginKeyProbe) that captures a
// just-registered passkey's PRF output, since registration alone doesn't
// reliably return it across every browser/authenticator. Called by
// webauthn.js's registerPasskey() immediately after register/finish
// succeeds — a failure here must never be treated as the passkey
// registration itself failing (see handleWebAuthnRegisterProbeFinish).
func (h *Handler) handleWebAuthnRegisterProbeBegin(w http.ResponseWriter, r *http.Request) {
	username, ceremonyKey, _, credentialID, ok := h.parseProbeRequest(w, r)
	if !ok {
		return
	}

	assertion, err := h.webauthn.BeginKeyProbe(r.Context(), username, ceremonyKey, credentialID)
	if err != nil {
		h.logger.Warn("webauthn: begin key probe failed", "username", username, "error", err)
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}
	writeJSON(w, assertion)
}

// handleWebAuthnRegisterProbeFinish completes the probe ceremony and hands
// its PRF output (if any) to KeyringService.AddPasskeySlot. Any failure here
// — the ceremony itself, or AddPasskeySlot's ErrNotUnlocked when the caller's
// session never bootstrapped a master key — is reported to the client so
// the profile screen can say "this passkey works for login, but not for
// encrypted-data unlock," but is deliberately never surfaced as the passkey
// registration having failed; that already succeeded before this ever ran.
func (h *Handler) handleWebAuthnRegisterProbeFinish(w http.ResponseWriter, r *http.Request) {
	username, ceremonyKey, body, _, ok := h.parseProbeRequest(w, r)
	if !ok {
		return
	}

	// credentialID here is the id FinishKeyProbe cryptographically validated
	// the assertion against — deliberately not the credentialId field
	// parseProbeRequest also decoded from this same request body, which is
	// just an unverified client claim. Binding the keyring slot to that
	// claim instead would let a client that legitimately owns credential A
	// mislabel its PRF output as belonging to credential B.
	credentialID, prfOutput, err := h.webauthn.FinishKeyProbe(r.Context(), username, ceremonyKey, body)
	if err != nil {
		h.logger.Warn("webauthn: finish key probe failed", "username", username, "error", err)
		http.Error(w, "key probe failed: "+err.Error(), http.StatusBadRequest)
		return
	}
	if len(prfOutput) == 0 {
		http.Error(w, "authenticator did not return a prf result", http.StatusUnprocessableEntity)
		return
	}

	if h.keyring == nil {
		w.WriteHeader(http.StatusNoContent)
		return
	}
	if err := h.keyring.AddPasskeySlot(r.Context(), username, credentialID, prfOutput); err != nil {
		h.logger.Warn("keyring: add passkey slot failed", "username", username, "error", err)
		http.Error(w, "add passkey slot failed: "+err.Error(), http.StatusConflict)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

File: internal/webui/static/js/webauthn.js
// Vanilla-JS glue between the browser's native WebAuthn API
// (navigator.credentials — no library needed on this side at all) and
// Miranda's /api/webauthn/* endpoints (internal/webui/webauthn.go). Shared
// by the profile screen (registering a new passkey) and the login page
// (passwordless/usernameless sign-in). Also owns the browser-local "last
// login method" hint (see LOGIN_METHOD_KEY below) that login.js uses to
// decide which method to lead with.
//
// The server's JSON (github.com/go-webauthn/webauthn/protocol) encodes byte
// fields — challenge, user.id, credential ids — as base64url strings;
// navigator.credentials.create()/get() need real ArrayBuffers instead, and
// the reverse conversion applies to what they return. These are the only
// two directions that need converting.

function base64urlToBuffer(b64url) {
  const padded = b64url + "=".repeat((4 - (b64url.length % 4)) % 4);
  const base64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  const raw = atob(base64);
  const bytes = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) bytes[i] = raw.charCodeAt(i);
  return bytes.buffer;
}

function bufferToBase64url(buffer) {
  const bytes = new Uint8Array(buffer);
  let str = "";
  for (const b of bytes) str += String.fromCharCode(b);
  return btoa(str).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function prepareCreationOptions(publicKey) {
  publicKey.challenge = base64urlToBuffer(publicKey.challenge);
  publicKey.user.id = base64urlToBuffer(publicKey.user.id);
  if (publicKey.excludeCredentials) {
    publicKey.excludeCredentials = publicKey.excludeCredentials.map((c) => ({ ...c, id: base64urlToBuffer(c.id) }));
  }
  return publicKey;
}

// PRF_SALT_B64URL is a fixed, non-secret, application-specific salt for the
// WebAuthn PRF extension's eval.first input — SHA-256("miranda:master-key:v1"),
// base64url-encoded (matches internal/keyring's docs/encryption.md). It
// doesn't need to be secret or per-user: PRF's security comes from the
// authenticator's own private key material, not this salt, which only
// domain-separates this specific derivation (like HKDF's "info" parameter).
// Requested client-side rather than round-tripped from the server to avoid
// a real encoding footgun — protocol.AuthenticationExtensions is untyped
// JSON, and a raw []byte value there would marshal as *standard* base64,
// not the base64url every other byte field in this protocol uses.
const PRF_SALT_B64URL = "1KUP_WqBdJvtc05WhVLE6ehg5_f950wDU5j-mzlvEnk";

function prepareRequestOptions(publicKey) {
  publicKey.challenge = base64urlToBuffer(publicKey.challenge);
  if (publicKey.allowCredentials) {
    publicKey.allowCredentials = publicKey.allowCredentials.map((c) => ({ ...c, id: base64urlToBuffer(c.id) }));
  }
  // Requested on every assertion (login and the post-registration PRF
  // probe below) — harmless no-op when the authenticator/browser doesn't
  // support PRF, or when internal/keyring isn't enabled server-side (the
  // output just goes unused in that case).
  publicKey.extensions = { ...(publicKey.extensions || {}), prf: { eval: { first: base64urlToBuffer(PRF_SALT_B64URL) } } };
  return publicKey;
}

// credentialToJSON mirrors the shape internal/webauthn's Service expects on
// finish — the same PublicKeyCredential JSON shape the WebAuthn spec itself
// defines, whether this came from create() (registration) or get() (login).
function credentialToJSON(credential) {
  const response = credential.response;
  const extensionResults = credential.getClientExtensionResults ? credential.getClientExtensionResults() : {};
  // The PRF extension's eval results arrive as raw ArrayBuffers, which
  // JSON.stringify silently serializes as "{}" (no enumerable own
  // properties) — encode them to base64url first, the same convention
  // every other binary field in this file already follows, or the server
  // never sees any bytes at all.
  if (extensionResults.prf && extensionResults.prf.results) {
    const results = extensionResults.prf.results;
    extensionResults.prf.results = {
      first: bufferToBase64url(results.first),
      ...(results.second ? { second: bufferToBase64url(results.second) } : {}),
    };
  }
  const out = {
    id: credential.id,
    rawId: bufferToBase64url(credential.rawId),
    type: credential.type,
    clientExtensionResults: extensionResults,
  };

  if (response.attestationObject) {
    out.response = {
      attestationObject: bufferToBase64url(response.attestationObject),
      clientDataJSON: bufferToBase64url(response.clientDataJSON),
      transports: response.getTransports ? response.getTransports() : [],
    };
  } else {
    out.response = {
      authenticatorData: bufferToBase64url(response.authenticatorData),
      clientDataJSON: bufferToBase64url(response.clientDataJSON),
      signature: bufferToBase64url(response.signature),
    };
    if (response.userHandle) out.response.userHandle = bufferToBase64url(response.userHandle);
  }
  return out;
}

/** Whether this browser/context supports WebAuthn at all — false on any
 * insecure (non-HTTPS, non-localhost) origin, which is exactly when the
 * passkey UI must stay hidden (see config.WebAuthnConfig's doc comment). */
export function isSupported() {
  return typeof window.PublicKeyCredential !== "undefined";
}

// LOGIN_METHOD_KEY backs a browser-local (not account-local) memory of
// which method last succeeded (or was last attempted, for "password" — see
// login.js's submit handler) from this browser — login.js's only way to
// guess, on the anonymous /login page, which method to lead with.
// Discoverable/usernameless login means the server can't tell who's about
// to sign in until the ceremony finishes, so there's no per-account signal
// to read here at all; a shared household device can carry one family
// member's choice into another member's visit, but that only costs them
// one click on the toggle button, never a dead end.
const LOGIN_METHOD_KEY = "miranda-last-login-method";

/** "passkey", "password", or null (no preference recorded yet) — see
 * LOGIN_METHOD_KEY. */
export function getLastLoginMethod() {
  try {
    return localStorage.getItem(LOGIN_METHOD_KEY);
  } catch {
    return null;
  }
}

// Best-effort: storage can throw in restricted contexts (e.g. Safari
// private browsing's quota), and the memory is a UX nicety, not required
// for correctness.
function setLastLoginMethod(method) {
  try {
    localStorage.setItem(LOGIN_METHOD_KEY, method);
  } catch {
    /* see doc comment above */
  }
}

/** Records a successful (or attempted — see login.js) passkey login, or a
 * successful passkey registration (profile screen), as the last-used
 * method. */
export function rememberPasskeyLogin() {
  setLastLoginMethod("passkey");
}

/** Records a password login attempt (login.js's form submit handler) as
 * the last-used method — see that handler for why an attempt, not just a
 * confirmed success, is what's recorded. */
export function rememberPasswordLogin() {
  setLastLoginMethod("password");
}

/** Clears the remembered method if it's "passkey" — called when the
 * profile screen finds this account has no passkeys left registered, since
 * that specific memory is now definitely stale for this account (leaves a
 * remembered "password" alone, which is still accurate either way). */
export function forgetPasskeyLogin() {
  try {
    if (localStorage.getItem(LOGIN_METHOD_KEY) === "passkey") {
      localStorage.removeItem(LOGIN_METHOD_KEY);
    }
  } catch {
    /* see setLastLoginMethod's doc comment */
  }
}

/** Runs the follow-up assertion ceremony that captures a just-registered
 * credential's PRF output for internal/keyring (see
 * internal/webauthn.Service.BeginKeyProbe's doc comment for why
 * registration alone can't reliably return it) — called by registerPasskey
 * immediately after registration succeeds. Deliberately fails soft: any
 * error here (no keyring configured server-side, the authenticator doesn't
 * support PRF, the ceremony itself fails) must never be treated as the
 * passkey registration having failed, since that already succeeded. Callers
 * that want to tell the user the outcome can inspect the resolved boolean.
 */
async function probeForEncryptionKey(credentialId) {
  try {
    const beginRes = await fetch("/api/webauthn/register/probe-begin", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ credentialId }),
    });
    if (!beginRes.ok) return false; // route not registered (keyring disabled) or begin failed
    const options = await beginRes.json();
    prepareRequestOptions(options.publicKey);

    const credential = await navigator.credentials.get({ publicKey: options.publicKey });
    const body = credentialToJSON(credential);
    body.credentialId = credentialId;

    const finishRes = await fetch("/api/webauthn/register/probe-finish", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    return finishRes.ok;
  } catch {
    return false;
  }
}

/** Registers a new passkey for the logged-in user (profile screen).
 * Resolves to {id, nickname, ...} plus one extra field this function adds,
 * encryptionKeyEnabled, telling the caller whether probeForEncryptionKey
 * successfully added a wrapped-key slot for this passkey — purely
 * informational, never affects whether registration itself succeeded. */
export async function registerPasskey(nickname) {
  const beginRes = await fetch("/api/webauthn/register/begin", { method: "POST" });
  if (!beginRes.ok) throw new Error(`begin failed: ${beginRes.status}`);
  const options = await beginRes.json();
  prepareCreationOptions(options.publicKey);

  const credential = await navigator.credentials.create({ publicKey: options.publicKey });
  const body = credentialToJSON(credential);
  body.nickname = nickname;

  const finishRes = await fetch("/api/webauthn/register/finish", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!finishRes.ok) throw new Error(`registration failed: ${finishRes.status} ${await finishRes.text()}`);
  rememberPasskeyLogin();
  const info = await finishRes.json();

  info.encryptionKeyEnabled = info.id ? await probeForEncryptionKey(info.id) : false;
  return info;
}

/** Drives a passwordless, usernameless login from the login page's
 * biometric button. Resolves to the JSON {redirect} on success. */
export async function loginWithPasskey() {
  const beginRes = await fetch("/api/webauthn/login/begin", { method: "POST" });
  if (!beginRes.ok) throw new Error(`begin failed: ${beginRes.status}`);
  const options = await beginRes.json();
  const ceremonyId = options.ceremonyId;
  prepareRequestOptions(options.publicKey);

  const credential = await navigator.credentials.get({ publicKey: options.publicKey });
  const body = credentialToJSON(credential);
  body.ceremonyId = ceremonyId;

  const finishRes = await fetch("/api/webauthn/login/finish", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!finishRes.ok) throw new Error(`login failed: ${finishRes.status}`);
  rememberPasskeyLogin();
  return finishRes.json();
}

export async function listCredentials() {
  const res = await fetch("/api/webauthn/credentials");
  if (!res.ok) throw new Error(`list failed: ${res.status}`);
  return res.json();
}

export async function deleteCredential(id) {
  const res = await fetch(`/api/webauthn/credentials/${encodeURIComponent(id)}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`delete failed: ${res.status}`);
}
